librz/core/cmd/cmd_print.c

```diff
@@ -7198,3 +7198,674 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_swift_handler, RZ_LA
 CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_yara_handler, RZ_LANG_BYTE_ARRAY_YARA);
 #undef CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL
 #undef CMD_PRINT_BYTE_ARRAY_HANDLER_ENDIAN
+
+static void disassembly_n_instructions_as_table(RzTable *t, RzCore *core, int n_instrs) {
+	ut8 buffer[256];
+	rz_table_set_columnsf(t, "snssssss", "name", "addr", "bytes", "disasm", "comment", "esil", "refs", "xrefs");
+	const int minopsz = 1;
+	const int options = RZ_ANALYSIS_OP_MASK_BASIC | RZ_ANALYSIS_OP_MASK_HINT | RZ_ANALYSIS_OP_MASK_DISASM | RZ_ANALYSIS_OP_MASK_ESIL;
+	ut64 offset = core->offset;
+	for (int i = 0; i < n_instrs; i++) {
+		RzAnalysisOp *op = rz_core_analysis_op(core, offset, options);
+		if (!op || op->size < 1) {
+			i += minopsz;
+			offset += minopsz;
+			continue;
+		}
+		const char *comment = rz_meta_get_string(core->analysis, RZ_META_TYPE_COMMENT, offset);
+		if (!comment) {
+			comment = "";
+		}
+		rz_io_read_at(core->io, offset, buffer, RZ_MIN(op->size, sizeof(buffer)));
+		char *bytes = rz_hex_bin2strdup(buffer, op->size);
+		RzFlagItem *flag = rz_flag_get_i(core->flags, offset);
+		char *function_name = flag ? flag->name : "";
+		const char *esil = RZ_STRBUF_SAFEGET(&op->esil);
+		char *refs = __op_refs(core, op, 0);
+		char *xrefs = __op_refs(core, op, 1);
+		rz_table_add_rowf(t, "sXssssss", function_name, offset, bytes, op->mnemonic, comment, esil, refs, xrefs);
+		free(bytes);
+		free(xrefs);
+		free(refs);
+		offset += op->size;
+		rz_analysis_op_free(op);
+	}
+}
+
+RZ_IPI RzCmdStatus rz_cmd_disassembly_n_instructions_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state) {
+	ut32 old_blocksize = core->blocksize;
+	ut64 old_offset = core->offset;
+
+	st64 parsed = argc > 1 ? (st64)rz_num_math(core->num, argv[1]) : 0;
+	if (parsed > ST16_MAX || parsed < ST16_MIN) {
+		RZ_LOG_ERROR("the number of instructions is too big (%d < n_instrs < %d).\n", ST16_MAX, ST16_MIN);
+		return RZ_CMD_STATUS_ERROR;
+	}
+	int n_instrs = parsed;
+	if (n_instrs < 0) {
+		ut64 new_offset = old_offset;
+		if (!rz_core_prevop_addr(core, old_offset, -n_instrs, &new_offset)) {
+			new_offset = rz_core_prevop_addr_force(core, old_offset, -n_instrs);
+		}
+		ut32 new_blocksize = new_offset - old_blocksize;
+		if (new_blocksize > old_blocksize) {
+			rz_core_block_size(core, new_blocksize);
+		}
+		rz_core_seek(core, new_offset, true);
+	}
+
+	switch (state->mode) {
+	case RZ_OUTPUT_MODE_STANDARD:
+		rz_core_print_disasm(core->print, core, core->offset, core->block, core->blocksize, RZ_ABS(n_instrs), 0, 1, false, NULL, NULL);
+		break;
+	case RZ_OUTPUT_MODE_TABLE:
+		disassembly_n_instructions_as_table(state->d.t, core, RZ_ABS(n_instrs));
+		break;
+	case RZ_OUTPUT_MODE_JSON:
+		rz_cmd_state_output_array_start(state);
+		rz_core_print_disasm_json(core, core->offset, core->block, core->blocksize, RZ_ABS(n_instrs), state->d.pj);
```

Comment:
It's a bit strange having the different parameters for the `rz_core_print_disasm_json()` with `rz_core_print_disasm()`. I think it should be one function, at the very least dispatching the calls, and accept `RzCmdStateOutput` instead.

Reply:
both functions needs to be rewritten. it's a big mess and i'm not going to change them in this PR.

Reply:
Maybe we should open an issue to track this refactor.